sophia/src/triple/stream/_error.rs

```diff
@@ -0,0 +1,214 @@
+{
+	/// Converts the ``StreamError` into its `SourceError`.
+	/// 
+	/// _Note_ this also works if `SinkError` is `Infallible`.
+    pub fn into_source_err(self) -> SourceErr {
+        match self {
+            Source(err) => err,
+            Sink(err) => err.into(),
+        }
+    }
+}
+
+impl<SourceErr, SinkErr> StreamError<SourceErr, SinkErr>
+where
+    SourceErr: 'static + Error + Into<SinkErr>,
+    SinkErr: 'static + Error,
+{
+	/// Converts the ``StreamError` into its `SinkError`.
+	/// 
+	/// _Note_ this also works if `SourceError` is `Infallible`.
+    pub fn into_sink_err(self) -> SinkErr {
+        match self {
+            Source(err) => err.into(),
+            Sink(err) => err,
+        }
+    }
+}
+
+impl<SinkErr> StreamError<Infallible, SinkErr>
+where
+    SinkErr: 'static + Error,
+{
+    /// Converts a `StreamError` from an infallible `Source`.
+    pub fn with_fallible_source<SourceErr>(self) -> StreamError<SourceErr, SinkErr>
+    where
+        SourceErr: 'static + Error,
+    {
+        match self {
+            Source(_) => unreachable!(),
+            Sink(err) => Sink(err),
+        }
+    }
+}
+
+impl<SourceErr> StreamError<SourceErr, Infallible>
+where
+    SourceErr: 'static + Error,
+{
+    /// Converts a `StreamError` from an infallible `Sink`.
+    pub fn with_fallible_sink<SinkErr>(self) -> StreamError<SourceErr, SinkErr>
+    where
+        SinkErr: 'static + Error,
+    {
+        match self {
+            Source(err) => Source(err),
+            Sink(_) => unreachable!(),
+        }
+    }
+}
+
+impl<Err> StreamError<Err, Err>
+where
+    Err: 'static + Error,
+{
+    /// Converts the `StreamError` into its underlying error if `Source` and
+    /// `Sink` errors are of the same type.
+    pub fn into_inner(self) -> Err {
+        match self {
+            Source(err) | Sink(err) => err,
+        }
+    }
+}
+
+impl<SourceErr, SinkErr> fmt::Display for StreamError<SourceErr, SinkErr>
+where
+    SourceErr: 'static + Error,
+    SinkErr: 'static + Error,
+{
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        use self::StreamError::*;
+
+        match self {
+            Source(err) => write!(f, "Source failed: {}", err),
+            Sink(err) => write!(f, "Sink failed: {}", err),
+        }
+    }
+}
+
+impl<SourceErr, SinkErr> Error for StreamError<SourceErr, SinkErr>
+where
+    SourceErr: 'static + Error,
+    SinkErr: 'static + Error,
+{
+    fn source(&self) -> Option<&(dyn Error + 'static)> {
+        match self {
+            Source(err) => Some(err),
+            Sink(err) => Some(err),
+        }
+    }
+}
+
+/// Extension trait for results to convienently convert to `StreamError`.
+///
+/// The counter part is [`SinkResultExt`](trait.SinkResultExt.html)
+pub trait SourceResultExt<T, SourceErr>
+where
+    SourceErr: 'static + Error,
+{
+    fn map_source_err<SinkErr>(self) -> Result<T, StreamError<SourceErr, SinkErr>>
+    where
+        SinkErr: 'static + Error;
+}
```

Comment:
I'm really not sure we need this (nor the following) extension trait.
`my_result.map_source_err()` could be written `my_result.map_err(SourceError)`,
which is much more idiomatic and hardly more verbose.